# Zamanian Lab Manual v 2.0

## I. Lab Mission and Philosophy

### A. Research Mission
The central ambition of our laboratory is to combine molecular biology, genetics, and computational approaches to improve our understanding of helminth (parasitic worm) biology and our ability to treat globally important human and animal parasitic infections. This includes discovering new drug targets, elucidating mechanisms of drug resistance, and developing new tools for parasite manipulation. Our current work involves the study of mosquito-borne and soil-transmitted nematode parasites, as well as snail (vector) biology in relation to schistosome control. To complement some of our efforts, we exploit the experimental tractability of the model nematode *Caenorhabditis elegans*. We strive to place all of our work in the context of a growing appreciation for and understanding of parasite genetic diversity.

### B. Philosophy

We look to recruit individuals who share our curiosity and who wish to challenge themselves as they grow their scientific independence. We welcome students and trainees from a variety of backgrounds and with different career ambitions. I take my role as a mentor and advisor very seriously and it is therefore important to me that all members of the lab: (1) remain excited about the science we do, (2) challenge themselves to improve their base of knowledge and technical skills, and (3) engage in meaningful self-reflection about their progress and goals.

Borrowing from Federico Ardila-Mantilla, we build our lab upon these axioms:

- **Axiom 1.** Scientific “talent” is distributed equally among different groups, irrespective of geographic, demographic, and economic boundaries.

- **Axiom 2.** Everyone can have joyful, meaningful, and empowering scientific experiences.

- **Axiom 3.** Science is a powerful, malleable tool that can be shaped and used differently by various communities to serve their needs.

- **Axiom 4.** Everyone deserves to be treated with dignity and respect.

### C. Mentorship

We will hold weekly lab meetings as a springboard for open and critical scientific exchange. I will also schedule regular individual meetings with every member of the lab so that we can discuss research progress and whatever else that may arise. Career development is a multi-faceted endeavor. Where I am positioned to do so, I will identify or help create opportunities that relate to the career aspirations of individual lab members. I will work with trainees to help them hone their scientific writing and communication skills. This will occur through the drafting of manuscripts, as well as the attendance of relevant conferences. Once lab members have a firm grasp of their primary research projects, I will often encourage them to move out of their comfort zones. For many, this may mean achieving fluency in a programming language (e.g., R or Python) useful to their research and future. Lab members should also seek to make use of the expertise of the broader lab and university community. In general, your resources should include a wide network of peers and mentors. Foster relationships with others in your same position and those who are willing to offer advice. It is critical in science to foster your own support network of people you trust, are inspired by, and that have expertise you don’t yet have.


### D. Lab Citizenship

My commitment to your success in research and professional growth is coupled to certain expectations. I expect all members of the lab to take ownership of their projects. I expect members of the lab to lend their help and expertise to one another and understand that our laboratory is a team with common scientific goals. I expect everyone to display good lab citizenship. We place value in an open, collaborative, and safe scientific environment. We expect all members of the lab to be good lab citizens and to follow basic lab rules.

    1. Be helpful to others - this is a team effort!
    2. Carry out your lab tasks in a timely manner.
    3. Follow rules for ordering and inventorying items.
    4. Keep your items confined to assigned spaces.
    5. Do not borrow things without asking.
    6. Clean up after yourselves and put things back in assigned locations.
    7. Be here during normal work hours so that you can maximize interactions
    8. Follow all rules when using facilities or resources outside of our lab.
    9. Plan the use of parasites many weeks in advance.
    10. Make sure the lab is properly shut down and communicate before leaving.

*Lab dress code:* For your safety, wear long pants or skirts and closed toed shoes, and tie back long hair. Do not wear shorts, short skirts, sandals, loose clothing, or dangling jewelry. Throughout your scientific career, you will be professionally interacting with people from diverse backgrounds and with diverse beliefs. In the interests of maintaining an inclusive and welcoming lab environment, I ask that you refrain from wearing articles of clothing that are likely to be perceived as hostile to another person’s heritage, ethnicity, gender, religious beliefs, or sexuality.   

## II. Communication and Calendar

### A. Slack

All members of the lab will be asked to install **[Slack](<http://www.slack.com>)** on their computers and phones and will receive an invitation to join the official lab Slack team [zamanian-lab.slack.com](<http://zamanian-lab.slack.com>). Slack will provide a basis for rapid individual and group communication on all research and lab-related matters.

### B. Google Calendar

An official lab Google calendar will be shared and updated with lab meeting times, relevant departmental events, and one-on-one meeting times. Lab members who will be out of town should notify me in advance and mark those dates on the lab calendar so that other lab members are aware should anything arise. Approved vacation days and other time off should also be entered into the lab calendar.

## III. Research Documentation

### A. Physical Lab Notebooks

All members of the lab should still record the details of their daily research activities in the form of a traditional lab notebook with dated entries. This is a requirement for federally-funded research. Lab notebooks cannot leave the laboratory and must be turned in at the conclusion of your tenure in the lab.

### B. Electronic Lab Notebooks

The laboratory uses **[Benchling](<http://www.benchling.com>)** as a shared electronic lab notebook (ELN) and central repository for all molecular biology efforts. Lab members will receive an invitation to join the ZamanianLab Benchling organization page. Electronic summaries of work done and data generated (gels and printouts) should be regularly annotated and recorded into your defined Benchling lab notebook project space. This means that your electronic notes should include the minimal set of notes required to repeat any molecular biology experiment that you have performed. Do not let too much time elapse before you convert your physical notes into electronic form, as it is easier to clear up ambiguities when experiments are fresh in memory. This should develop into a habit. Benchling also provides tools for the planning of molecular cloning tasks and the design of molecular biology reagents.

### C. Lab Electronic Storage

All grad students and postdocs should install the **[Box Drive](<http://www.box.com>)** application and receive an invitation to edit the ZamanianLab shared Box folder. All students and staff of the University of Wisconsin-Madison receive Box cloud storage at no cost through their university e-mail accounts. The shared folder will be used to store important lab data, scripts, and documents. **DO NOT** delete shared folders while your local computer is set to sync as this will delete data across all devices. Do not use spaces in file names, unnecessarily long file names, or vague file naming systems that obscure contents. The primary storage folders are described below:

- ``/Data/`` This folder should be used to neatly preserve original files and raw data output from various lab and campus instruments (e.g., subfolders ``/Gels/``, ``/Sanger/``, ``/qPCR/``), common assays and projects with small data footprints (e.g., subfolders ``/Ce_assays/`` and ``/RNAi_experiments/``), as well as minimally processed 'big' data outputs (e.g., subfolder ``/RNAseq/``). Other subfolders will be created to represent different project-specific data types where necessary, with strict guidelines for file nomenclature. These folders have clear organization and naming conventions that need to followed.

- ``/LabMembers/`` Individual lab folders are provided for members of the lab to store their research-related documents. These individual folders should be used to preserve processed data, protocols in the process of optimization, writing (manuscripts, fellowships, and grants), and any computer programs or scripts written to carry out preliminary analyses.

The uses of other Box folders are described in subsequent sections.

### D. Lab Meetings

Lab meetings will be held once a week. Every member of the lab should have an updated presentation file (dated ``YYYYMMDD``) in their personal ``/LabMeetings/`` subfolder. This presentation should contain your overall project goals, divided into sub-aims, as well as data and timelines for specific experiments completed and planned for the coming weeks. You should be prepared to present your progress in this short format every week when called upon. This is a space for all members of the lab to contribute to troubleshooting and to engage in constructive and healthy scientific criticism.

### E. One-on-One Meetings

I will hold weekly one-on-meetings with every graduate student and postdoc, except when I'm traveling or have other commitments that I cannot sideline. You should come prepared with organized summaries of the status of your project(s), to share data you’ve produced, and to outline your next set of planned experiments. Each trainee and I will share a **Google Doc** to keep track of weekly and long-term goals. We will use this time to troubleshoot and constructively dissect these data and plans. This is also an opportunity for us to discuss general issues and your career development goals.

## IV. Lab Tasks

Every member of the lab will be involved in scheduled lab maintenance in some capacity. This will include things like defined cleaning activities and making common buffers and reagents. Roles are divided between undergraduates and other lab staff: **[Undergrad Tasks](protocols/Tasks/UG_Tasks/UG_Tasks.md)** and **[Staff Tasks](protocols/Tasks/Staff_Tasks/Staff_Tasks.md)**.


## V. General Lab Organization

### A. Lab Inventory

The locations of **all reagents and consumables** will be stored on **[Quartzy](<http://www.quartzy.com>)**. Lab members will be expected to return ALL items to their assigned locations after use. It is the responsibility of lab members to make sure the Quartzy inventory is up-to-date. Additional rules are outlined in subsequent sections for inventory of plasmids, bacteria, worm strains, oligos, and parasite materials.

### B. Lab Ordering

Lab order requests should be made through Quartzy. Lab orders can be placed after review and approval. Orders can be placed using either the UW **[MDS](<https://mds.bussvc.wisc.edu/order/shopper_lookup.asp>)** system or lab purchasing card (P-card). **It is the responsibility of every member of the lab to communicate when we are nearing the end of necessary supplies or reagents so that orders can be placed in a timely manner.** Except for routine items (e.g., tips, tubes, and gloves), you should message the PI on Slack prior to placing an order on Quartzy to receive approval. Each request should be associated with a funding source in the Quartzy order form. Orders will be fulfilled on designated days of each week. **When items arrive, they should be immediately marked as received, physically dated, and associated with a location on Quartzy.**

**P-card ordering:** Items will occasionally have to be ordered with the lab P-card. P-cards are also often used for conference registration. I expect lab trainees to process these orders and to place a pdf receipt (appropriately named) of their order in ``/LabOrdering/PCARD_receipts`` for the lab manager to process.

### C. Physical Organization

**Space Assignments:** Benches and Pipettes will be assigned to each grad student and postdoc. Undergraduates should only use the spaces and pipettes of their lab mentor.

**Storage:** Each graduate student and postdoc will have their own rack spaces in the -20C and -80C freezers for personal storage of materials related to individual experiments. Each person is responsible for maintaining proper labeling and notes of their stored items. There are designated storage spaces for lab and commercial plasmids, various bacterials stocks, *C. elegans* strain stocks, and parasite materials across lab freezers.

### D. Lab Operations and Waste Management

Place dirty glassware in assigned sink locations (Rms 223 and 230). Place empty pipette boxes in assigned sink location (Rm 230). Place worm boxes for baking in assigned location (Rm 223). Empty shipping boxes should be placed on the ground outside the lab.


## VI. Lab Stocks

Locations and descriptions of all lab worm strains, plasmids, bacteria, and oligos should be stored and updated on Quartzy and in defined spreadsheets where specified. The following protocols and naming conventions must be followed, and lab members are responsible for checking the lab database to avoid naming conflicts prior to placing any new orders and assigning new names. Bacteria, worm strains, and plasmids procured from external sources and with existing names should retain those names.

### A. Oligo Stocks

Use ``IDT_oligos.xls`` in ``/LabOrdering/Oligos/`` to enter oligo details for ordering. Check Quartzy to ensure you are using the next oligo number available. Provide complete information (Species, Oligo Pairing, Molecular Target, Amplicon Length, and Tm) within the spreadsheet order form. All oligo types (DNA primers, gRNAs) should be assigned an ID with the same naming convention. Orders will be placed by a designated lab member.

```
Oligo Name: oMZ####
Storage box labels (-20C): "Oligo Stocks #"
* Quartzy Update (Type = 'Oligo or Primer'):
    Item Name (=Oligo Name)
    Sequence
    Technical Details (Species, Target, Tm, Pairing)
    Location
```

**Primers:** Make 100 uM lab stocks of each primer oligo and store in Oligo Stocks box (-20C). Add 10*Y (uL) volume to lyophilized primer, where Y is nmol yield to make 100 uM stocks. Keep 10 uM (1:10 of stock) dilutions of primers in your own personal boxes for personal use. Re-order the original stock primer before it runs out.

### B. Plasmids (Lab and Commercial)                

Lab Plasmids are plasmids created or designed by our lab for long-term use and storage. Commercial (external) Plasmids are plasmids we have received as gifts or ordered from vendors.

```
Plasmid Name: pMZ#### (lab) or Original Name
Storage box labels (-80C & -20C):
    "Lab Plasmids #" or "Commercial Plasmids #"
Tube Counts: 1 copy (-80C) + 1 or more copies (-20C)
* Tube Labels
    Plasmid Name
    Concentration
    Date Frozen
* Quartzy Update (Type = 'Plasmids'):
    Item Name (=Plasmid Name)
    Amount In Stock (units)
    Location
    Date Frozen
    Resistance Marker
    Use (e.g., "gene X RNAi" or "C. elegans transgene")
```

Use the stock(s) in -20C for all molecular biology purposes. When -20C stocks have been depleted, transfer the backup stock from -80C to -20C, and inoculate liquid media with the frozen bacterial stock to replace the plasmid stock in -80C. All intermediary vectors using for cloning that are unlikely to be used for other purposes should be stored in personal boxes until no longer needed. **Plasmids note:** All completed (inventoried) plasmid sequences should be deposited into the designated lab plasmid Benchling project. Additionally, plasmid sequences and image maps should be exported to the Box folder ```/LabInventory/Plasmid_Export/```.

### C. Transformed Bacteria

```
Storage box labels (-80C & N2):
  "Lab Plasmid Bacterial Stock #" and "Commercial Plasmid Bacterial Stock #"
Tube Counts: 2 copies (-80C) + 1 copy (N2)
* Tube Labels
    Plasmid Name
    Bacteria (eg., DH5a)
    Date Frozen
* Quartzy Update (Type = 'Transformed Bacteria'):
    Item Name (=Plasmid Name)
    Amount In Stock (units)
    Location
    Bacteria (eg., DH5a)
    Date Frozen
    Plasmid (=Plasmid name)
    Resistance Marker
```

### D. Feeding (*C. elegans*) Bacteria

Boxes specific to each bacteria (e.g., OP50, HB101, NA22) used for *C. elegans* feeding should be stored in the designated area of the -80C, noted with how many tubes of each bacteria per box in Quartzy. **Replace all bacterial stocks before using up the -80C stock.**

```
Storage box labels (-80C):
  "Bacteria Name"
Tube Counts: Variable
* Tube Labels
    Bacteria (e.g., OP50)
    Date Frozen
* Quartzy Update (Type = 'Feeding Bacteria'):
    Item Name (=Bacteria Name)
    Amount In Stock (units)
    Location
    Date Frozen
```


### E. Competent Bacteria

Boxes specific to each competent bacteria (DH5a, NE5a, JM109) should be stored in the designated area of the -80C, noted with how many tubes of each bacteria per box in Quartzy. **Replace all bacterial stocks before using up the -80C stock.**

```
Storage box labels (-80C):
  "Bacteria Name"
Tube Counts: Variable
* Tube Labels
    Bacteria (e.g., DH5a)
    Date Frozen
* Quartzy Update (Type = 'Competent Bacteria'):
    Item Name (=Bacteria Name)
    Location
    Date Frozen
```

## VI. *Caenorhabditis elegans* Strains

### A. Strain Creation

Only the lab manager (Kathy) is allowed to use the microinjection apparatus to create transgenic strains. It will be the responsibility of any person who wishes for a strain to be made to:

    1. Coordinate with Kathy at least two weeks in advance
    2. Provide necessary background strain for injection (if not N2)
    3. Provide injection mixes in the Plasmids box in the top freezer in room 223
      - Mark the tube with the date and all plasmids included in the mix
      - Including any fluorescent or phenotypic marker that should be followed
    4. Provide all required information in the inventory spreadsheet (see below)

It is your responsibility to monitor injected lines and select for stable lines and confirming genotypes (through PCR and other approaches). Provide Kathy with the F1 plates to chunk from and cryogenically preserve once you have decide to keep any transgenic lines.


### B. Strain Inventory

**Lab strain designation: ZAM#**  
**Lab allele designation: maz#**

*C. elegans* strains we receive from CGC, as gifts from other labs, or that we create in the lab will be frozen by the lab manager. Five tubes will be frozen at once, with one of those tubes frozen in the Liquid Nitrogen Tank. Strains must be associated with COMPLETE genotypic information (e.g., Transgenes, Alleles, Strain Background) on Quartzy AND in the **C. elegans Strain Database** linked in **[Lab Sheets](labsheets.md)**.

### C. Strain Maintenance

Maintain strains in your assigned areas of the 15/20C incubators, or above your bench if incubating at ambient temperature. Use assigned colors to label your plate boxes. Wrap old plates and dispose of contaminated plates regularly. **Your contaminated plates (bacteria, fungi, or mites) can jeopardize the experiments of others!** Request chunks or frozen strains from the lab manager with adequate notice. Strains should be re-thawed or chunked at least once every six months to avoid mutation accumulation. Update the Rm 230 whiteboard with the strains you are actively maintaining. Boxes should be baked before using them to propagate new strains.

## VII. Parasite and Insectary Operations

### A. Parasite Delivery Schedule

We generate parasites through local infections and receive regular shipments of filarial parasites (*B. malayi*, *B. pahangi*, and *D. immitis*) from the NIH FR3 repository. The Lab Manager will place external orders for many months ahead and the expected scheduling of all parasite stage and materials will be updated on the **Parasite Schedule Sheet** linked in [Lab Sheets](labsheets.md). Keep track of both local and external parasite orders. Shipments do not always arrive as expected, but **you should have a detailed list of future parasite needs on the spreadsheet**. I will send out a FedEx confirmation once I receive it (typically one day in advance) in the #Parasite_Planning Slack channel.

### B. Storage of General Parasite Materials

General use (non-experimental) parasite materials should be catalogued and stored in designated racks/boxes. There are individual boxes for *B. malayi*, *B. pahangi, and D. immitis* tissues and RNA in the -80C freezer. There is a single box for storage of parasite gDNA in the -20C freezer. The contents of all samples in these boxes should be inventoried for each species in the appropriate tab of the **Parasite Tissue and Nucleic Acid Inventory** sheet linked in [Lab Sheets](labsheets.md). The spreadsheet should be immediately updated after placing any new tubes in these boxes. This catalog of materials allows us to plan qPCR, RNA-seq and other experiments.

```
* Tube Labels (combine the following into a single string)
  1) Species (First letter of genus + first letter of species)
  2) Date (YYYYMMDD)
  3) Sample Type (T=tissue, R=RNA, CD=cDNA, GD=gDNA)
  4) Number ID (T###)
  Example: Bm20190530.T001
```

### C. Parasite Infections (Insectary)

The Bartholomay Lab insectary is a powerful resource that comes with major responsibilities. All members of the lab should follow established safety protocols and basic lab etiquette when using this space. No member of the lab should carry out insectary operations and procedures without approval from both PIs. Detailed protocols for insectary operations, filarial parasite infection, and L3 isolation can be found on our [Protocols](protocols/protocols.md) page. Mosquito strains available for parasite infection:

- ***Ae. aegypti LVP*** strain: used for *B. malayi* and *B. pahangi* infections
- ***Ae. aegypti SD*** strain: used for *D. immitis* infections

Strains must be ordered in advance for infection, accounting for potential die offs and typical L3 yields. Place Orders as instructed in the [Insectary Protocol](protocols/Parasitic_Nematodes/Insectary/Insectary.md) and update the **Parasite Schedule Sheet** linked in [Lab Sheets](labsheets.md) to associate parasite batches with mosquito infections

```
Mosquito Carton Labels
  Mosquito spp.:
  Parasite spp.:
  Parasite Source:
  Date Infected:
  Titer:
  Extraction Date:
```

## VIII. Writing and Reference Management

To streamline collaborative writing and/or editing of manuscripts and fellowship applications, lab members should create [Google Docs](<http://docs.google.com>) to share documents with me and others involved in the writing goal. [Paperpile](<https://paperpile.com/>) will be used to manage references. Paperpile is compatible with Google Docs. A lab license will be made available to those who require it. For those who prefer LaTeX over traditional word processing, [OverLeaf](<http://www.overleaf.com>) is also an acceptable alternative to Google Docs. All final data, figures, scripts, and text associated with manuscript submissions should be organized within ``/Manuscripts/`` and in a manuscript-specific GitHub project.


## IX. Lab Safety

All lab members must take the appropriate safety training before starting any lab work and to use appropriate PPE at all times. Lab biosafety protocols and procedures will also be provided in binder and electronic form (``/LabSafety/``). In the event of emergencies involving accident or injury, call 911 to be routed to the campus police.

- Training required for everyone

    - "Biosafety Required Training" ([EHS](<https://ehs.wisc.edu/biosafety-training/>))

- Animal-related training for IACUC-approved work

    - "Animal User Orientation" ([RARC](<https://www.rarc.wisc.edu/services/training/getting_started.html>))
    - "Safety for Personnel with Animal Contact" ([EHS](<https://ehs.wisc.edu/animal-safety-training/>))
    - "Animal Contact Risk Questionnaire (ACRQ)" ([UHS](<https://www.uhs.wisc.edu/eoh/animal-contact/>))
    - Additional 1-in-1 Training Offered through RARC (e.g., Mouse Training)

- For work with human cell lines and potentially infectious materials
    - "Biosafety 102: Bloodborne Pathogens for Laboratory and Research" ([EHS](<https://ehs.wisc.edu/biosafety-training/>))

**E-mail both the Bartholomay and Zamanian lab** list-servs (bartlab at lists.wisc.edu and zamanianlab at lists.wisc.edu) in advance of BSL2 work in the insectary space. Put up BSL-2 signs on doors and include location, time, and reminder for PPE.

Example e-mail: *"I will be performing bulk parasite extractions from infected LVP mosquitoes TOMORROW (27/11/18). I will be performing the extractions in the screened-in work space inside the insectary (room 322). The parasites I will be extracting are Brugia pahangi L3s, and are classified as a BSL-2 organism. I will be working from 9:30AM - 12:30 PM, and will have signs posted on the outside of the screened-in area. If you need to enter this area during this time, please mind the PPE requirements specified on the posted BSL-2 signs. If this is a conflict/problem for anybody, please contact me as soon as possible. Thank you."*

## X. Lab Onboarding Checklist

1. Room key and building card access form
2. Office space, bench, and pipette assignment
3. Electronic accounts (will use UW email)
    - Quartzy
    - Benchling
    - Box
    - Google Calendar
    - Google Group (for shared sheets)
    - Paperpile
    - GitHub
    - CHTC
4. Biosafety Training
5. Provide photo for lab website
